### PR TITLE
connectivity: Add test for HostPort support

### PR DESCRIPTION
**Pull request by gandro**
This PR adds an additional scenario to the `no-policies` test which checks connectivity to pods with a HostPort. HostPort support in Cilium can be achieved either via CNI chaining with the portmap plugin, or via Kube Proxy Replacement. I have validated this PR with both setups.

There are some changes needed to the overall code, such as support for conditional scenarios and support for alternative destination addresses in flow validation (needed for KPR HostPort). The newly introduced scenario is only run if the HostPort feature is detected.

*Review per commit*

### CODE REVIEW

**Comment by gandro** [comment]
**Open question**: Note that in CI we currently don't run against a cluster with hostport support. I wonder if we should enable it e.g. in the Kind workflow to make sure this code doesn't break accidentally?

**Comment by gandro** [comment]
CI failures look legitimate. Moving back to draft.

**Comment by tklauser** [comment]
> **Open question**: Note that in CI we currently don't run against a cluster with hostport support. I wonder if we should enable it e.g. in the Kind workflow to make sure this code doesn't break accidentally?

IMO it'd make sense to extend the kind workflow (or add a new workflow) covering hostport support.

**Comment by gandro** [comment]
> > **Open question**: Note that in CI we currently don't run against a cluster with hostport support. I wonder if we should enable it e.g. in the Kind workflow to make sure this code doesn't break accidentally?
> 
> IMO it'd make sense to extend the kind workflow (or add a new workflow) covering hostport support.

I added a commit, but turns out the CNI chaining feature detection does not work on v1.11, because `cilium status` does not report chaining status :thinking:

**Comment by tklauser** [comment]
> > > **Open question**: Note that in CI we currently don't run against a cluster with hostport support. I wonder if we should enable it e.g. in the Kind workflow to make sure this code doesn't break accidentally?
> > 
> > 
> > IMO it'd make sense to extend the kind workflow (or add a new workflow) covering hostport support.
> 
> I added a commit, but turns out the CNI chaining feature detection does not work on v1.11, because `cilium status` does not report chaining status thinking

Oh, we're still running CI with v1.11.6. I think it'd be fine to update CI to use v1.12.0 on `master` branch in this repo.

**Comment by gandro** [comment]
Added a commit which adds support for Cilium v1.11 feature detection as well.

**Comment by gandro** [comment]
Working in CI now as well: https://github.com/cilium/cilium-cli/runs/7669029226?check_suite_focus=true

```
[-] Scenario [no-policies/pod-to-hostport]
  [.] Action [no-policies/pod-to-hostport/curl-0: cilium-test/client-6488dcf5d4-g2z28 (10.244.0.131) -> cilium-test/echo-other-node-57c4b6df47-2nxxd (172.18.0.3:40000)]
  🐛 Connected to Hubble node(s) [kind-chart-testing/chart-testing-control-plane kind-chart-testing/chart-testing-worker]
  📄 Following flows...
  🐛 Executing command [curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null http://172.18.0.3:40000]
  📄 Validating flows for peer cilium-test/client-6488dcf5d4-g2z28
  🐛 Validating 20 flows against 1 requirements
  ✅ SYN found at 0
  ✅ SYN-ACK found at 2
  ✅ FIN found at 10
  ✅ L3/L4 Drop not found
  ✅ RST not found
  ✅ Flow validation successful for peer cilium-test/client-6488dcf5d4-g2z28 (first: 0, last: 10, matched: 3)
  
  [.] Action [no-policies/pod-to-hostport/curl-1: cilium-test/client-6488dcf5d4-g2z28 (10.244.0.131) -> cilium-test/echo-same-node-54f79b9d4d-8xzts (172.18.0.2:40000)]
  🐛 Connected to Hubble node(s) [kind-chart-testing/chart-testing-control-plane kind-chart-testing/chart-testing-worker]
  📄 Following flows...
  🐛 Executing command [curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null http://172.18.0.2:40000]
  📄 Validating flows for peer cilium-test/client-6488dcf5d4-g2z28
  🐛 Validating 20 flows against 1 requirements
  ✅ SYN found at 0
  ✅ SYN-ACK found at 2
  ✅ FIN found at 14
  ✅ L3/L4 Drop not found
  ✅ RST not found
  ✅ Flow validation successful for peer cilium-test/client-6488dcf5d4-g2z28 (first: 0, last: 14, matched: 3)
  
  [.] Action [no-policies/pod-to-hostport/curl-2: cilium-test/client2-6dd75b74c6-j8zqg (10.244.0.150) -> cilium-test/echo-other-node-57c4b6df47-2nxxd (172.18.0.3:40000)]
  🐛 Connected to Hubble node(s) [kind-chart-testing/chart-testing-worker kind-chart-testing/chart-testing-control-plane]
  📄 Following flows...
  🐛 Executing command [curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null http://172.18.0.3:40000]
  📄 Validating flows for peer cilium-test/client2-6dd75b74c6-j8zqg
  🐛 Validating 20 flows against 1 requirements
  ✅ SYN found at 0
  ✅ SYN-ACK found at 2
  ✅ FIN found at 14
  ✅ L3/L4 Drop not found
  ✅ RST not found
  ✅ Flow validation successful for peer cilium-test/client2-6dd75b74c6-j8zqg (first: 0, last: 14, matched: 3)
  
  [.] Action [no-policies/pod-to-hostport/curl-3: cilium-test/client2-6dd75b74c6-j8zqg (10.244.0.150) -> cilium-test/echo-same-node-54f79b9d4d-8xzts (172.18.0.2:40000)]
  🐛 Connected to Hubble node(s) [kind-chart-testing/chart-testing-control-plane kind-chart-testing/chart-testing-worker]
  📄 Following flows...
  🐛 Executing command [curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null http://172.18.0.2:40000]
  📄 Validating flows for peer cilium-test/client2-6dd75b74c6-j8zqg
  🐛 Validating 20 flows against 1 requirements
  ✅ SYN found at 0
  ✅ SYN-ACK found at 2
  ✅ FIN found at 14
  ✅ L3/L4 Drop not found
  ✅ RST not found
  ✅ Flow validation successful for peer cilium-test/client2-6dd75b74c6-j8zqg (first: 0, last: 14, matched: 3)
```

File: connectivity/check/test.go [hunk 1]
```diff
@@ -178,7 +191,7 @@ func (t *Test) Run(ctx context.Context) error {
 
 		sn := t.scenarioName(s)
```
Comment (tklauser) [inline connectivity/check/test.go:192]: Nit: probably can move this inside the `t.Logf` below now that it's the only user of `sn`.

**Review by tklauser — APPROVED** [review]
Nice work!

Two small nits inline, but probably not worth addressing if there are no other changes requested.

File: connectivity/check/features.go [hunk 2]
```diff
@@ -213,6 +215,19 @@ func (ct *ConnectivityTest) extractFeaturesFromCiliumStatus(ctx context.Context,
 	return nil
 }
 
+// deriveFeatures derives additional features based on the status of other features
+func (ct *ConnectivityTest) deriveFeatures(result FeatureSet) error {
```
[the remark this answers is not in the document]
Comment (gandro) [reply connectivity/check/features.go:239]: Good suggestion, will implement.

**Comment by gandro** [comment]
Addressed Tobias' feedback.

**Review by nbusseneau — APPROVED** [review]
LGTM.

**Pull request merged**